exchange.go

```diff
@@ -193,14 +193,43 @@ func (ex *exchange) Run(ctx context.Context, inp, out chan Dataset) (err error)
 func (ex *exchange) send(data Dataset) error {
 	switch ex.Type {
 	case scatter:
-		return ex.encodeNext(data)
+		return ex.encodeParallelToPeers(data)
 	case partition:
 		return ex.encodePartition(data)
 	default:
 		return ex.encodeAll(data)
 	}
 }
 
+func (ex *exchange) encodeParallelToPeers(data Dataset) error {
+	amountOfPeers := len(ex.encs)
+	modulo := data.Len() % amountOfPeers
+	batchSize := data.Len() / amountOfPeers
+	start := 0
+	var err error
+
+	// send remainder batch size, i.e. send the modulo data size
+	for i := 0; i < modulo; i++ {
+		end := start + batchSize + 1
+		err = ex.encodeNext(data.Slice(start, end).(Dataset))
```

Comment:
you don't really need to cast here. encodeNext gets interface :)
```suggestion
		err = ex.encodeNext(data.Slice(start, end))
```